File: json-schema/research/format_uri/native_probes/native.py
#!/usr/bin/env python3
"""Python NATIVE URI-parser probe (urllib.parse). For each input, reports what
urllib.parse.urlsplit thinks. urllib almost NEVER raises; a naive stdlib
format:uri check would treat "has a scheme" as valid absolute. We report
`bool(scheme)` as the "valid absolute" verdict, which is what a naive user does.

Emits JSON Lines: {"id","engine":"python-native","valid":bool,"detail":str}
Run: python3 native.py ../native_inputs.json
"""
import json
import sys
from urllib.parse import urlsplit


def main() -> None:
    path = sys.argv[1] if len(sys.argv) > 1 else "../native_inputs.json"
    with open(path, encoding="utf-8") as fh:
        corpus = json.load(fh)
    out = sys.stdout
    for inp in corpus["inputs"]:
        valid = False
        detail = ""
        try:
            parts = urlsplit(inp["value"])
            if parts.scheme:
                valid = True
                detail = f"scheme={parts.scheme} netloc={parts.netloc!r}"
            else:
                detail = "no scheme"
        except ValueError as e:
            detail = f"error: {e}"
        out.write(
            json.dumps(
                {"id": inp["id"], "engine": "python-native", "valid": valid, "detail": detail},
                ensure_ascii=False,
            )
            + "\n"
        )


if __name__ == "__main__":
    main()

File: json-schema/research/format_uri/native_probes/native.rb
# encoding: UTF-8
# frozen_string_literal: true
#
# Ruby NATIVE URI-parser probe. Ruby's `URI.parse` / `URI.regexp` uses an
# RFC-2396-derived grammar (with RFC3986 parser available). We report:
#   URI.parse succeeds AND the result is a URI::Generic with a non-nil scheme
#     AND it is absolute? => "valid absolute".
# `URI.parse` raises URI::InvalidURIError on many malformed inputs.
#
# Emits JSON Lines: {"id","engine":"ruby-native","valid":bool,"detail":string}
# Run: ruby native.rb ../native_inputs.json
require "json"
require "uri"

def main
  path = ARGV[0] || "../native_inputs.json"
  corpus = JSON.parse(File.read(path, encoding: "UTF-8"))
  corpus["inputs"].each do |inp|
    valid = false
    detail = ""
    begin
      u = URI.parse(inp["value"])
      if u.absolute?
        valid = true
        detail = "scheme=#{u.scheme}"
      else
        detail = "parsed but not absolute (scheme=#{u.scheme.inspect})"
      end
    rescue URI::InvalidURIError => e
      detail = "error: #{e.message}"
    rescue StandardError => e
      detail = "error(#{e.class}): #{e.message}"
    end
    $stdout.puts JSON.generate("id" => inp["id"], "engine" => "ruby-native",
                               "valid" => valid, "detail" => detail)
  end
end

main if __FILE__ == $PROGRAM_NAME
